File: frontend/src/types/subscription.ts
export type PlanTier = 'starter' | 'professional' | 'enterprise'

export type BillingCycle = 'monthly' | 'annual'

export type SubscriptionStatus = 'trial' | 'active' | 'canceled' | 'past_due'

export interface SubscriptionPlan {
  id: string
  tier: string
  name: string
  description?: string
  monthlyPrice: number
  annualPrice: number
  maxUsers?: number
  maxProjects?: number
  maxStorageGb?: number
  isActive: boolean
  createdAt: string
  updatedAt: string
}

export interface Subscription {
  id: string
  organizationId: string
  planId: string
  billingCycle: BillingCycle
  status: SubscriptionStatus
  trialEndsAt?: string
  currentPeriodStart: string
  currentPeriodEnd: string
  canceledAt?: string
  stripeSubscriptionId?: string
  payplusSubscriptionId?: string
  createdAt: string
  updatedAt: string
  plan?: SubscriptionPlan
}

export interface SubscriptionCreate {
  organizationId: string
  planId: string
  billingCycle?: BillingCycle
  trialDays?: number
}

export interface SubscriptionUpgrade {
  planId: string
  billingCycle?: BillingCycle
}

export interface SubscriptionCancel {
  immediate?: boolean
}

File: backend/app/models/subscription.py
import uuid
from datetime import datetime
from enum import Enum

from sqlalchemy import DateTime, ForeignKey, Integer, Numeric, String, Boolean
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base
from app.utils import utcnow


class PlanTier(str, Enum):
    STARTER = "starter"
    PROFESSIONAL = "professional"
    ENTERPRISE = "enterprise"


class SubscriptionPlan(Base):
    __tablename__ = "subscription_plans"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tier: Mapped[str] = mapped_column(String(50), nullable=False, unique=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(String(500), nullable=True)
    monthly_price: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    annual_price: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    max_users: Mapped[int | None] = mapped_column(Integer, nullable=True)
    max_projects: Mapped[int | None] = mapped_column(Integer, nullable=True)
    max_storage_gb: Mapped[int | None] = mapped_column(Integer, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: utcnow())
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: utcnow(), onupdate=lambda: utcnow())

    subscriptions = relationship("Subscription", back_populates="plan")


class Subscription(Base):
    __tablename__ = "subscriptions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    organization_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("organizations.id", ondelete="CASCADE"), index=True, unique=True)
    plan_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("subscription_plans.id"), index=True)
    billing_cycle: Mapped[str] = mapped_column(String(20), nullable=False)  # "monthly" or "annual"
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="trial")  # trial, active, canceled, past_due
    trial_ends_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    current_period_start: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=lambda: utcnow())
    current_period_end: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    canceled_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    stripe_subscription_id: Mapped[str | None] = mapped_column(String(255), nullable=True, unique=True)
    payplus_subscription_id: Mapped[str | None] = mapped_column(String(255), nullable=True, unique=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: utcnow())
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: utcnow(), onupdate=lambda: utcnow())

    organization = relationship("Organization")
    plan = relationship("SubscriptionPlan", back_populates="subscriptions")
